lib/device.c

```diff
@@ -475,45 +671,25 @@ static int handle_del_key_cmd(SOCKET sock, struct sync_device *data)
 	return sync_del_key(data->tracks[track], row);
 }
 
-int sync_tcp_connect(struct sync_device *d, const char *host, unsigned short port)
+int sync_server_setup_tcp(struct sync_device *d, const char *host, unsigned short port)
 {
-	int i;
-	if (d->sock != INVALID_SOCKET)
-		closesocket(d->sock);
-
-	d->sock = server_connect(host, port);
-	if (d->sock == INVALID_SOCKET)
-		return -1;
-
-	for (i = 0; i < (int)d->num_tracks; ++i) {
-		free(d->tracks[i]->keys);
-		d->tracks[i]->keys = NULL;
-		d->tracks[i]->num_keys = 0;
-	}
-
-	for (i = 0; i < (int)d->num_tracks; ++i) {
-		if (fetch_track_data(d, d->tracks[i])) {
-			closesocket(d->sock);
-			d->sock = INVALID_SOCKET;
-			return -1;
-		}
-	}
-	return 0;
-}
-
-int sync_connect(struct sync_device *d, const char *host, unsigned short port)
-{
-	return sync_tcp_connect(d, host, port);
+	return server_setup(d, host, port);
 }
 
 int sync_update(struct sync_device *d, int row, struct sync_cb *cb,
     void *cb_param)
 {
-	if (d->sock == INVALID_SOCKET)
-		return -1;
+	int	r;
+
+	r = server_ready(d);
+	if (r <= 0)
+		return -1;	/* TODO: it would be better to propagate out the server_ready()
+				 * return values from sync_update() so the demo can know if it's
+				 * currently EINPROGRESS vs. proper errors
+				 */
```

Comment:
To be a bit clearer about my previous comment: I don't think we should automatically reconnect in `sync_update()`. It would IMO be better if reconnection was an explicit operation from the calling code.

One reason is that some users *might* want to handle a disconnection differently than what the example does. It also seems like a logically different operation...

Reply:
But then what do `sync_tcp_connect()` and `sync_reconnect()` return when it's just a non-blocking connect that's been initiated but not completed?  Do we return a success status as if the connect succeeded?  And if the connection is still in-progress when `sync_update()` is called, is it returning a success so it doesn't trigger a reconnect?

My problem with changing the meaning of the return values without changing the function names in the interests of supporting non-blocking connects is programs written for the old behavior will still compile successfully, but the semantics have changed.

I deliberately deprecated the old `sync_tcp_connect()` so compiles against the new code will fail and force developers to understand why, then update the `sync_update()` return handling accordingly.  It's not like the changes would be substantial.

I'm ok with making the reconnect explicit with the introduction of `sync_reconnect()`, but I think we still need to deprecate `sync_tcp_connect()` to both break builds *and* better reflect what's going on.

How about this:

`sync_server_setup_tcp()`: performs blocking resolution of specified host, registers the addr+port on the sync device as the server, does not initiate a connect.

`sync_connect()`: initiates potentially non-blocking connect of the device to the registered server addr+port, may block if platform doesn't support non-blocking connects

`sync_update()`: returns different values for SUCCESS, CONNECT_IN_PROGRESS, DISCONNECTED

Is there still a need for `sync_reconnect()` in the above?  You'd just use `sync_connect()` with the registered server addr+port...

WDYT